### DISCUSSION TITLE
Mixing micropython projects

### DISCUSSION THREAD

**Discussion by kc64** [BrianPugh/belay#170] [Q&A] [answered]
We have multiple devices performing different functions using micropython on the RP2040. One device has analog input code and timer code. Another device uses the PIO state machine and analog input code. Another device uses the PIO state machine and SPI bus.

We would like to maintain the analog input code and the PIO state machine code in separate code bases and “import” them so we don’t break DRY. It turns out this is rather tricky to do. If I open a device and want to apply the analog code and the PIO code with the belay setup method, there are device conflicts. The decorator does not seem to work with already established/connected devices.

I should also point out that it is a strong project goal to not install any code on the device itself, but rather to load all of it at runtime.

Any ideas?

**Comment by BrianPugh** [comment]
You should only create the actual `belay.Device` object in your main project repo. What's the desire of keeping the micropython code in a separate repo? Generally, the micropython code in a Belay project is minimal/simple. I'll need more context on your project!

**Reply by kc64** [reply]
I'll just take the PIO case. We have 230+ lines of PIO state machine code in our project which is used as a subproject in other projects.

Case 1: I combine that PIO code with some SPI bus devices and I2C devices to make a signal-processing controller.
Case 2: I combine that PIO code with some discrete GPIO processing code and RS232 serial comm code to make a pulse generator.

I'd rather keep the PIO code in one project and pull it in for both cases. In the CPython world, I might make a mixin class and subclass the designs for the two cases or import the PIO module and build on it. I'm looking for a similar model for the belay instance.

**Reply by BrianPugh** [reply]
gotcha, here are my thoughts.

1. Lets say your micropython PIO code is in the repo `micropython-pio`
2. Use some mechanism to get those code files into your main project (submoduling, fetching, something!).
3. In your Device subclass, you can do something like:

```
from belay import Device

class MyDevice(belay.Device):
    def __pre_autoinit__(self):
        # This writes your micropython pio code to the microcontroller's flash. Alternatively, use ``self.sync``
        self.sync_dependencies("relative-path-to-micropython-pio")

    @Device.setup(autoinit=True)
    def setup():
        # this runs in micropython land, imports the files synced from above.
        import micropython_pio
        # Also setup your i2c and other long-lived objects here
        i2c = I2C(...)
        my_i2c_chip = Max722823(i2c, ...)
    
    @Device.task
    def read(self):
        # use the objects from setup
        return my_i2c_chip.read()
        
```

Is this helpful at all? or am I missing the issue?

**Reply by BrianPugh** [reply]
here's a project I did a while ago that might be a useful reference:

https://github.com/BrianPugh/magnetometer/blob/main/magnetometer/sensors/base.py

**Reply by kc64** [reply]
I think this may just come down to the last clause of my original post: “I should also point out that it is a strong project goal to not install any code on the device itself, but rather to load all of it at runtime.”

Maybe this is just not doable with my exclusion.

**Reply by BrianPugh** [reply, marked as answer]
Here's another solution that might work without writing to the microcontroller's filesystem:

```
from belay import Device
from pathlib import Path

class MyDevice(Device):
    def __pre_autoinit__(self):
        micropython_code = Path("micropython-code.py").read_text()
        self(micropython_code)  # Executes the contents of "micropython-code.py" on device in a global context.
        
    @Device.task
    def read(self):
        return my_instantiated_pio_thing.get_some_value()
```

A similar technique is used internally (see the private `Device._exec_snippet` method) to send over functions and other startup code for belay.

**Comment by kc64** [comment]
Using this helpful example, I have a working solution that I think will serve well in distributing this software.

Belay is an impressive accomplishment. The support that you provide is equally impressive. Many thanks.

**Reply by BrianPugh** [reply]
Thank you! Please dont hesitate to make more post/issues if you run into any trouble!
